File: Data_structures/Linked List/linked_list.cpp
#include <iostream>
#include <cmath>
#include <algorithm>
using namespace std;

#define sci(x) scanf("%d", &x)
#define print(x) printf("%d ", x)

struct Node{
    int data;
    Node *nextNode;
};

Node* addToList(Node *root, int data){
    if(root == NULL){
        Node *newNode = new Node;
        newNode->data = data;
        newNode->nextNode = NULL;
        return newNode;
    }

    root->nextNode = addToList(root->nextNode, data);
    return root;
}

void printList(Node *head){
    if(head == NULL){
        return;
    }
    print(head->data);
    printList(head->nextNode);
}

int main(){
    Node *headPtr = NULL;

    headPtr = addToList(headPtr, 2);
    headPtr = addToList(headPtr, 3);
    headPtr = addToList(headPtr, 4);
    headPtr = addToList(headPtr, 5);

    printList(headPtr);

    return 0;
}

File: Data_structures/Linked List/linked_list.py
class Node():
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList():
    def __init__(self):
        self.head = None


    def add(self, data):
        if self.head == None:
            self.head = Node(data)
            return
        
        newNode = Node(data)
        curNode = self.head
        # prevNode = self.head

        while(curNode.next != None):
            curNode = curNode.next
        
        curNode.next = newNode

    
    def printList(self):
        curNode = self.head

        while(curNode != None):
            print(curNode.data)
            curNode = curNode.next


if __name__ == '__main__':
    linked_list = LinkedList()
    linked_list.add(2)
    linked_list.add(3)
    linked_list.add(4)

    linked_list.printList()


